### PR TITLE
test(map): add unit tests for MapController layer reconciliation

### LINKED ISSUE #351: Add unit tests for map-controller.ts (layer reconciliation core)
## Problem

`packages/map/src/map-controller.ts` (~1918 lines) is the central layer reconciliation/orchestration module and is the highest-risk code to change, but it has no dedicated unit test. `layer-sync.ts` has a few tests (vector/raster/geoagent), but the controller orchestration is uncovered. The largest stateful UI surfaces (`StylePanel`, `AddDataDialog`, `TopToolbar`, `AttributeTable`, `DesktopShell`) also rely solely on the 8 Playwright smoke specs.

## Proposal

Add unit tests for `map-controller.ts` covering syncLayers reconciliation (add/remove/reorder/update, source vs layer changes) against a mocked MapLibre instance, then extend coverage to the largest untested panels.

### PR DESCRIPTION
**Pull request by giswqs**
## Summary

Closes #351.

`packages/map/src/map-controller.ts` is the central layer reconciliation/orchestration module and the highest-risk code to change, but it had no dedicated unit test. This adds `tests/map-controller.test.ts`, exercising the controller against a stateful fake MapLibre instance (no DOM/WebGL needed — a stub map plus `styleReady` is injected so `syncLayers` and the camera/identify helpers run in plain Node, matching the existing layer-sync test style).

## Coverage

**`syncLayers` reconciliation:**
- no-op until the style is ready
- adds a layer's source + style layer, tracking `layerIds`/`syncedLayers`
- removes native source + layers when a layer drops out
- reorders style layers via `moveLayer` (not teardown) when order changes
- in-place `setData` on geojson updates vs. **source recreation** when clustering toggles
- visibility toggle applied as a layout property
- repaint on a paint-style change
- drop-and-replace teardown when a layer id is replaced

**Basemap controls:** hide/show basemap style layers, exclude user style layers from the basemap set, scale opacity from the original paint value.

**Camera/query helpers:** `readView`, `readProjection`, `fitBounds` (degenerate point → flyTo, non-finite box rejected), `identifyFeatures`, `setStoryLayerOpacity` (write-through + clamping).

**Built-in control positions:** defaults and recording a position for a hidden control.

## Testing

- `npm run test:frontend` — 787 tests pass (21 new)
- `pre-commit run --files tests/map-controller.test.ts` — all hooks pass

## Summary by CodeRabbit

* **Tests**
  * Added comprehensive automated coverage for the map controller, including deterministic layer synchronization (adding, removing, reordering, and updating GeoJSON without recreating sources).
  * Verified basemap behavior (visibility toggling, opacity scaling) and correct scoping to user layers.
  * Covered camera and query helpers, including `fitBounds` edge cases, feature identification, and opacity updates with clamping/optional transitions.
  * Added checks for built-in control positioning and safe control add/remove behavior.

### CODE REVIEW

File: tests/map-controller.test.ts [hunk 1]
```diff
@@ -0,0 +1,559 @@
+import assert from "node:assert/strict";
+import { beforeEach, describe, it } from "node:test";
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/map-controller.test.ts]: Done in 6af9102 — dropped the unused `beforeEach` import.

File: tests/map-controller.test.ts [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+      calls.push({ method: "setLayoutProperty", args: [id, key, value] });
+    },
+    setLayerZoomRange: record("setLayerZoomRange"),
+    // Camera + query helpers used by the controller's public methods.
+    project: (lngLat: [number, number]) => ({ x: lngLat[0], y: lngLat[1] }),
+    queryRenderedFeatures: () => {
+      const features = pendingRenderedFeatures;
+      pendingRenderedFeatures = [];
+      return features;
+    },
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/map-controller.test.ts]: Done in 6af9102 — the fake `queryRenderedFeatures` now honors the `layers` filter and no longer clears the queue on read, so identifyFeatures' per-layer loop can't starve later layers. Added a scoped multi-layer identify test covering it.

File: tests/map-controller.test.ts [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+      pitch: 0,
+      bbox: [-120, 30, -80, 50],
+    });
+  });
+
+  it("normalizes the projection to globe/mercator", () => {
+    const { map } = makeFakeMap();
+    const controller = controllerWith(map);
+    assert.equal(controller.readProjection(), "mercator");
+  });
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/map-controller.test.ts:471]: Done in 6af9102 — added a test asserting a non-mercator (`globe`) projection normalizes to `"globe"`.

File: tests/map-controller.test.ts [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+      (c) =>
+        c.method === "setPaintProperty" &&
+        c.args[0] === circleId("a") &&
+        c.args[1] === "circle-opacity",
+    );
+    assert.ok(op);
+    assert.equal(op.args[2], 1);
+  });
+});
+
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/map-controller.test.ts:595]: Done in 6af9102 — added a `setStoryLayerOpacity("a", 0.5, 500)` test asserting the `circle-opacity-transition` paint property is set to `{ duration: 500 }`.

File: tests/map-controller.test.ts [hunk 5]
```diff
@@ -0,0 +1,559 @@
...
+    // geolocate defaults to hidden, so setting its position just records it.
+    const ok = controller.setBuiltInControlPosition("geolocate", "bottom-right");
+
+    assert.equal(ok, true);
+    assert.equal(
+      controller.getBuiltInControlPosition("geolocate"),
+      "bottom-right",
+    );
+  });
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/map-controller.test.ts:619]: Done in 6af9102 — added `addControl`/`removeControl` stubs to the fake map, plus tests for the controller's public `addControl` passthrough and `removeControl` error-swallowing. A full visible built-in reposition test can't run here: that path re-creates a real maplibregl control whose constructor needs a DOM (`window`), which `node --test` doesn't provide (verified: `new maplibregl.NavigationControl()` throws `window is not defined`). Noted in a comment at that spot.

File: tests/map-controller.test.ts [hunk 6]
```diff
@@ -0,0 +1,559 @@
...
+      if (spec) {
+        const layout = (spec.layout as Record<string, unknown>) ?? {};
+        layout[key] = value;
+        spec.layout = layout;
+      }
+      calls.push({ method: "setLayoutProperty", args: [id, key, value] });
+    },
+    setLayerZoomRange: record("setLayerZoomRange"),
+    // Camera + query helpers used by the controller's public methods.
+    project: (lngLat: [number, number]) => ({ x: lngLat[0], y: lngLat[1] }),
```
[the remark this answers is not in the document]
Comment (giswqs) [reply tests/map-controller.test.ts:139]: Acknowledged — leaving the `project` stub returning raw lng/lat as-is. No current test depends on projected pixel accuracy (identifyFeatures' point is passed straight to `queryRenderedFeatures`, which the fake matches by layer id, not pixel position), and you flagged this as low value. Happy to add a Mercator projection if a future coordinate-dependent test needs it.
